### PR TITLE
add initial version of VM and trace

### PR DESCRIPTION
**Pull request by yaanhyy**
1. add VM with instruction name.
2. add trace

### CODE REVIEW

File: processor/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+                    // let src_index = self.get_reg_index(ops.get(2).unwrap());
+                    let value = self.get_index_value(ops.get(2).unwrap());
+                    self.flag = self.registers[dst_index] == value.0;
+                    program.trace.insert_step(self.clk, self.pc as u32, self.fp[self.fp.len()-1] as u32, Instruction::EQ(Equal{ri: dst_index as u8, a: value.1}),
+                                              self.registers.clone(),self.flag, None);
+                    self.pc += 1;
+                }
+                "gt" => {
+                    debug!("opcode: gt");
+                    assert!(ops.len() == 3 ,  "gt params len is 2");
```
Comment (spartucus) [inline processor/src/lib.rs:88]:
```suggestion
                    assert!(ops.len() == 3 ,  "gt params len is 3");
```

**Comment by yaanhyy** [comment]
> Note that the instruction set needs to be changed in the following version :)

sure!

File: processor/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    pub fn execute(&mut self, program: &mut Program) -> Result<(), String> {
+        loop {
+            let ops_line =  program.instructions[self.pc as usize].trim();
+            let ops: Vec<&str> = ops_line.split(' ').collect();
+            let mut opcode = ops.get(0).unwrap().to_lowercase();
+            opcode = opcode.to_lowercase();
+            match opcode.as_str() {
+                "mov" => {
+                    debug!("opcode: mov");
+                    assert!(ops.len() == 3 ,  "movi params len is 2");
```
Comment (spartucus) [inline processor/src/lib.rs:66]:
```suggestion
                    assert!(ops.len() == 3 ,  "movi params len is 3");
```

File: processor/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                    let value = self.get_index_value(ops.get(2).unwrap());
+                    self.registers[dst_index] = value.0;
+
+                    program.trace.insert_step(self.clk, self.pc as u32, self.fp[self.fp.len()-1] as u32, Instruction::MOV(Mov{ri:dst_index as u8,a:value.1}),
+                                              self.registers.clone(),self.flag, None);
+                    self.pc += 1;
+                },
+                "eq" => {
+                    debug!("opcode: eq");
+                    assert!(ops.len() == 3 ,  "eq params len is 2");
```
Comment (spartucus) [inline processor/src/lib.rs:77]: ditto

File: processor/src/lib.rs (hunk 1 above)
Comment (spartucus) [inline processor/src/lib.rs:88]: ditto

File: processor/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                    // let src_index = self.get_reg_index(ops.get(2).unwrap());
+                    let value = self.get_index_value(ops.get(2).unwrap());
+                    self.flag = self.registers[dst_index] > value.0;
+                    program.trace.insert_step(self.clk, self.pc as u32, self.fp[self.fp.len()-1] as u32, Instruction::GT(Gt{ri: dst_index as u8, a: value.1}),
+                                              self.registers.clone(),self.flag, None);
+                    self.pc += 1;
+                }
+                "lt" => {
+                    debug!("opcode: lt");
+                    assert!(ops.len() == 3 ,  "lt params len is 2");
```
Comment (spartucus) [inline processor/src/lib.rs:99]: ditto

File: processor/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                    let value = self.get_index_value(ops.get(1).unwrap());
+                    program.trace.insert_step(self.clk, self.pc as u32, self.fp[self.fp.len()-1] as u32, Instruction::JMP(Jmp{a: value.1}),
+                                              self.registers.clone(),self.flag, None);
+                    self.fp.push(self.pc +1);
+                    self.pc = value.0;
+
+                }
+                "add" => {
+                    debug!("opcode: add");
+                    assert!(ops.len() == 4 ,  "add params len is 2");
```
Comment (spartucus) [inline processor/src/lib.rs:138]: ditto

File: processor/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+                    let op2_value = self.get_index_value(ops.get(3).unwrap());
+                    self.registers[dst_index] =  self.registers[op1_index] + op2_value.0;
+                    program.trace.insert_step(self.clk, self.pc as u32, self.fp[self.fp.len()-1] as u32,
+                                              Instruction::ADD(Add{ri: dst_index as u8, rj: op1_index as u8, a: op2_value.1}),
+                                              self.registers.clone(),self.flag, None);
+                    self.pc += 1;
+                }
+                "sub" => {
+                    debug!("opcode: sub");
+                    assert!(ops.len() == 4 ,  "sub params len is 2");
```
Comment (spartucus) [inline processor/src/lib.rs:150]: ditto

File: processor/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+                    let op2_value = self.get_index_value(ops.get(3).unwrap());
+                    self.registers[dst_index] = self.registers[op1_index] - op2_value.0;
+                    program.trace.insert_step(self.clk, self.pc as u32, self.fp[self.fp.len()-1] as u32,
+                                              Instruction::SUB(Sub{ri: dst_index as u8, rj: op1_index as u8, a: op2_value.1}),
+                                              self.registers.clone(),self.flag, None);
+                    self.pc += 1;
+                }
+                "mul" => {
+                    debug!("opcode: sub");
+                    assert!(ops.len() == 4 ,  "mul params len is 2");
```
Comment (spartucus) [inline processor/src/lib.rs:162]: ditto

File: processor/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+                    let op2_value = self.get_index_value(ops.get(3).unwrap());
+                    self.registers[dst_index] = self.registers[op1_index] * op2_value.0;
+                    program.trace.insert_step(self.clk, self.pc as u32, self.fp[self.fp.len()-1] as u32,
+                                              Instruction::MUL(Mul{ri: dst_index as u8, rj: op1_index as u8, a: op2_value.1}),
+                                              self.registers.clone(),self.flag, None);
+                    self.pc += 1;
+                }
+                "ret" => {
+                    debug!("opcode: ret");
+                    assert!(ops.len() == 2 ,  "ret params len is 1");
```
Comment (spartucus) [inline processor/src/lib.rs:174]: ditto

**Review by spartucus — CHANGES_REQUESTED** [review]
typos to be fixed

**Pull request closed without merging**